kiwi/builder/disk.py

```diff
@@ -444,144 +445,65 @@ def create_disk(self) -> Result:
-                self.boot_image.create_initrd(self.mbrid)
-
-            # create second stage metadata to system image
-            self._copy_first_boot_files_to_system_image()
-
-            self._write_bootloader_meta_data_to_system_image(
-                device_map, disk, system
-            )
-
-            self.mbrid.write_to_disk(
-                disk.storage_provider
-            )
-
-            # run pre sync script hook
-            if self.system_setup.script_exists(
-                defaults.PRE_DISK_SYNC_SCRIPT
-            ):
-                disk_system = SystemSetup(
-                    self.xml_state, self.root_dir
-                )
-                disk_system.call_pre_disk_script()
-
-            # syncing system data to disk image
-            self._sync_system_to_image(
-                device_map, system, system_boot, system_efi, system_spare,
-                system_custom_parts
-            )
-
-            # run post sync script hook
-            if self.system_setup.script_exists(
-                defaults.POST_DISK_SYNC_SCRIPT
-            ):
-                image_system = ImageSystem(
-                    device_map, self.root_dir,
-                    system.get_volumes() if self.volume_manager_name else {}
-                )
-                image_system.mount()
-                disk_system = SystemSetup(
-                    self.xml_state, image_system.mountpoint()
-                )
-                try:
-                    disk_system.call_disk_script()
-                finally:
-                    image_system.umount()
-
-            # install boot loader
-            self._install_bootloader(device_map, disk, system)
-
-            # set root filesystem properties
-            self._setup_property_root_is_readonly_snapshot(system)
-
-            Result.verify_image_size(
-                self.runtime_config.get_max_size_constraint(),
-                self.diskname
-            )
+                    ) as filesystem:
+                        if self.root_filesystem_embed_integrity_metadata:
+                            filesystem.create_on_device(
+                                label=self.disk_setup.get_root_label(),
+                                size=-defaults.DM_METADATA_OFFSET,
+                                unit=defaults.UNIT.byte
+                            )
+                        else:
+                            filesystem.create_on_device(
+                                label=self.disk_setup.get_root_label(),
+                            )
+                        system = filesystem
+
+        # sync system data, configure system, setup loader and initrd
+        try:
+            self._build_main_system(
+                device_map,
+                disk,
+                system,
+                system_boot,
+                system_efi,
+                system_spare,
+                system_custom_parts,
+                luks_root,
+                raid_root
+            )
```

Comment:
This does not look like a good idea. `__exit__()` of Filesystem is called in line 480, but you are reusing the Filesystem instance in the else case in line 486.

Reply:
See the other comments which should make this more clear. It's ok and wanted to use the instance of a filesystem after the context exit is done